mp2-v1/src/values_extraction/mod.rs

```diff
@@ -38,66 +41,165 @@ pub(crate) const OUTER_KEY_ID_PREFIX: &[u8] = b"OUT_KEY";
 
 pub(crate) const BLOCK_ID_DST: &[u8] = b"BLOCK_NUMBER";
 
+/// Storage slot information for generating the proof
+#[derive(Clone, Debug, Eq, PartialEq, Hash, Serialize, Deserialize)]
+pub struct StorageSlotInfo<const MAX_COLUMNS: usize, const MAX_FIELD_PER_EVM: usize> {
+    slot: StorageSlot,
+    metadata: MetadataGadget<MAX_COLUMNS, MAX_FIELD_PER_EVM>,
+    outer_key_id: F,
+    inner_key_id: F,
+}
+
+impl<const MAX_COLUMNS: usize, const MAX_FIELD_PER_EVM: usize>
+    StorageSlotInfo<MAX_COLUMNS, MAX_FIELD_PER_EVM>
+{
+    pub fn new(
+        slot: StorageSlot,
+        metadata: MetadataGadget<MAX_COLUMNS, MAX_FIELD_PER_EVM>,
+        outer_key_id: F,
+        inner_key_id: F,
+    ) -> Self {
+        Self {
+            slot,
+            metadata,
+            outer_key_id,
+            inner_key_id,
+        }
+    }
+
+    pub fn slot(&self) -> &StorageSlot {
+        &self.slot
+    }
+
+    pub fn metadata(&self) -> &MetadataGadget<MAX_COLUMNS, MAX_FIELD_PER_EVM> {
+        &self.metadata
+    }
+
+    pub fn outer_key_id(&self) -> F {
+        self.outer_key_id
+    }
+
+    pub fn inner_key_id(&self) -> F {
+        self.inner_key_id
+    }
+}
+
 pub fn identifier_block_column() -> u64 {
     let inputs: Vec<F> = BLOCK_ID_DST.to_fields();
     H::hash_no_pad(&inputs).elements[0].to_canonical_u64()
 }
 
-/// Calculate `id = Poseidon(slot || contract_address)[0]` for single variable.
+/// Compute identifier for single variable (value of Struct).
+/// `id = H(slot || evm_word || contract_address || chain_id)[0]`
 pub fn identifier_single_var_column(
     slot: u8,
+    evm_word: u32,
     contract_address: &Address,
     chain_id: u64,
     extra: Vec<u8>,
 ) -> u64 {
-    let fields = contract_address
-        .0
-        .to_vec()
-        .into_iter()
+    let inputs = once(slot)
+        .chain(evm_word.to_be_bytes())
+        .chain(contract_address.0.to_vec())
         .chain(chain_id.to_be_bytes())
-        .chain(extra.into_iter())
-        .collect::<Vec<u8>>()
-        .to_fields();
-
-    let inputs: Vec<_> = iter::once(GFp::from_canonical_u8(slot))
-        .chain(fields)
-        .collect();
+        .chain(extra)
+        .map(F::from_canonical_u8)
+        .collect_vec();
 
     H::hash_no_pad(&inputs).elements[0].to_canonical_u64()
 }
 
-/// Calculate `key_id = Poseidon(KEY || slot || contract_address)[0]` for mapping variable leaf.
+/// Compute key indetifier for mapping variable.
+/// `key_id = H(KEY || slot || evm_word || contract_address || chain_id)[0]`
 pub fn identifier_for_mapping_key_column(
     slot: u8,
+    evm_word: u32,
```

Comment:
`evm_word` shouldn't be necessary as input for mapping key identifiers, as there cannot be struct as mapping keys (and if there might be, they are not supported in circuits right now I think). So I think we could remove this input from all `identifier_for_mapping_key_*` methods (i.e., also the ones for outer and inner mapping keys). Wdyt?